Performance improvement when creating nodes

Llorx opened the pull request:
This patch will make the tree to be redrawn only one time when sorting is enabled and also will only create new nodes if needed, as addressed in https://github.com/vakata/jstree/issues/1406

vakata commented:
Thank you for the contribution but I'd like to keep the core code (`jstree.js`) unaware of plugin methods - calling `this.sort` from `jstree.js` is something I want to avoid. I will change the way plugins work in v.4, but until then I'd rather not merge this PR. I will give it some thought, so any ideas are welcome.

vakata commented:
As for optimizing `create_node` - v.4 already has that - there is no distinction between create and load, the underlying tree model is just that - a tree model, and why the nodes are added is irrelevant (loading, creating, etc).

Llorx commented:
Fine, looking towards v.4 with excitement. Any ETA? :P

I've modified the redraw so it now modified the `old` element instead of creating a new one, gaining much more performance. It checks if the element in the current array is different and does the according work (delete, create, change position).

I can remove the sort thingy if you want, but keep this changes as I think that they benefit the performance.

By the way, I'm still making changes. Will update here when I have them done, for you to check.

vakata commented:
@Llorx removing sort and adding the changes here would be great - thank you. I will review and merge after you are ready :) As for v.4 it is still a bit early to tell - I need to do a lot of work, so it all depends on how much free time I will have. I hope to have a beta out in May.

vakata commented:
As there is no recent activity - let's leave those changes in favor of v.4, which is making good progress and will have a separate branch next week.

VipinTiw commented:
Is this issue has been fixed and released in latest Jstree library.

vakata commented:
No, as I said - I am leaving the changes for v.4 which has not been released yet. 
If you are willing to remove the sort and apply all other changes you mentioned I will reopen?

VipinTiw commented:
Thanks Vakata, can you please share some work around solution on jsfiddle link.
Because still i am facing performance related issue with create node function.

vakata commented:
In the current v.3 version create_node should not be used to populate the tree - create_node is a user facing function - it should be called when the end user is creating a node via the UI. When creating the tree the load_node function is called which loads nodes according to the rules specified in core.data.
There is no work on a solution, except for this pull request - feel free to browse the code and pick what you like.

VipinTiw commented:
Thanks vakata, i tried with load_node and core.data but it did not work for me.
Actually i am loading original  jstree with json data format. 
  $("#" + _containerId).jstree({
            'core': {
                "check_callback": true,
                'data':treeViewData   // json data format
            }
        });
Tree  get loaded successfully. 
Now when i am clicking on any node, i want to add  child node only for clicked  node. In my case child node count is 3400. Earlier i had used create_node function , it was executing in a loop 3400 times.
It was taking nearly 30 sec to redraw tree.
Now i am using load_node function as below
        var newNode = {
            id: _selectedNodeId,         // selected or clicked folder
            state: {
                opened: true, // is the node open
                disabled: false, // is the node disabled
                selected: true  // is the node selected
            },
            children:data,  // array of  objects
        };
        jstree.load_node(newNode, function () {
        });
but nothing is happening.
Let me know where i am doing wrong.
TIA

VipinTiw commented:
Vakata waiting for your help. please provide your input on above issue.
TIA
